#link to github
https://github.com/nrossi2/SQL-Final-Project
    
#loom link
https://www.loom.com/share/6b5d447568a54015b8cf3000ab1537b4

In [2]:
#Connect to AWS database instance
%load_ext sql
%sql mysql://USERNAME:PASSWORD@HOST/DATABASE

'Connected: admin@Spotify_Data'

Question 1: What is the average popularity of the songs in the track table?

With the aim of the project surrounding the question of song popularity, it is good to start out by examining the average popularity of the songs we have availbale data on. This will reaffirm our confidence in our assumption that we will be analyzing that audio features of popular songs.

In [2]:
%%sql
SELECT AVG(track_popularity)
FROM tracks

 * mysql://admin:***@lmu-sql.cwmw1ha3dmhf.us-east-2.rds.amazonaws.com/Spotify_Data
1 rows affected.


[(Decimal('76.3462'),)]

The average popularity of the the songs from our tracks table is just above 76. Considering the range of this value is between 0 and 100, this is a relatively high figure. This reassures us that we will be working with tracks with an upwardly skewed value for popularity.

Question 2: What genres and artists are present in our data?

It is important to have knowledge of what exact artists and genres we will be running our analysis on. Certian artists or genres may be inherently higher in specific audio features. We want to reassure that we have diverse set of types of popular music. 

In [9]:
%%sql
SELECT artist_name, artist_genre
FROM artists

 * mysql://admin:***@lmu-sql.cwmw1ha3dmhf.us-east-2.rds.amazonaws.com/Spotify_Data
20 rows affected.


[('Maroon 5', 'pop'),
 ('Taylor Swift', 'dance pop'),
 ('Queen', 'glam rock'),
 ('Lady Gaga', 'dance pop'),
 ('Post Malone', 'dfw rap'),
 ('Mumford & Sons', 'modern folk rock'),
 ('Elton John', 'glam rock'),
 ('50 Cent', 'east coast hip hop'),
 ('Drake', 'canadian hip hop'),
 ('The Beatles', 'british invasion'),
 ('Coldplay', 'permanent wave'),
 ('Mariah Carey', 'dance pop'),
 ('DaBaby', 'nc hip hop'),
 ('Imagine Dragons', 'modern rock'),
 ('Lil Wayne', 'hip hop'),
 ('Kanye West', 'chicago rap'),
 ('Rihanna', 'barbadian pop'),
 ('Ariana Grande', 'dance pop'),
 ('Billie Eilish', 'electropop'),
 ('Jonas Brothers', 'boy band')]

This query provides us with the list of artists we will be looking through the data of. Before analyzing the songs in our data, it is important to have the knowledge of who created them. This will provide us with better context before we further sift through the dataset.

Question 3: How many songs are in the dataset?

It is helpful to have an idea of the scope of the data before proceeding.

In [25]:
%%sql
SELECT COUNT(*)
FROM tracks

 * mysql://admin:***@lmu-sql.cwmw1ha3dmhf.us-east-2.rds.amazonaws.com/Spotify_Data
1 rows affected.


[(182,)]

As we can see, we have a grand total of 182 songs present in our data. Although limited, this intentional in part. I wanted to conduct audio analysis of the most popular songs of Universal's most popular artists. I began by selecting 20 of their top artists and then selecting a collection of these artist's best performing work. This will provide us with a sample of the biggest commercial hits that Universal has in their collection.

Primary Question:
    
What audio features are most important in determining track popularity?

If Universal can determine, through quantative analysis, what audio features strongly indicate popularity, the company can then pursue artists who also score highly in these categories. It will provide them with stronger confidence that their signing efforts can be backed up by a foundation of numeric evaluation.

To first asnwer this question, I will write a series of queries that will evalaute the average popularity of tracks that score above average on each individual audio feature. These queries will also display the count of these tracks that score above average to provide an additional metric in which we can reconcile audio features and song performance. 

Queries will be structured utilzing both a JOIN and SUBQUERY. The JOIN will allow us to combine both the audio features and tracks table. This is important as the combined table will provide both the track_popularity column and each individual audio feature column. A SUBQUERY will be used as a means of comparison between each individual track's specific audio feature value and the overall average value for that audio feature.

In [19]:
%%sql
SELECT AVG(tracks.track_popularity) As track_popularity, COUNT(*) AS track_count
FROM tracks
JOIN audio_features
    ON tracks.track_id = audio_features.track_id
WHERE audio_features.danceability > 
    (
    SELECT AVG(danceability)
    FROM audio_features
    )

 * mysql://admin:***@lmu-sql.cwmw1ha3dmhf.us-east-2.rds.amazonaws.com/Spotify_Data
1 rows affected.


[(Decimal('76.9263'), 95)]

In [26]:
%%sql
SELECT AVG(tracks.track_popularity) As track_popularity, COUNT(*) AS track_count
FROM tracks
JOIN audio_features
    ON tracks.track_id = audio_features.track_id
WHERE audio_features.energy > 
    (
    SELECT AVG(energy)
    FROM audio_features
    )

 * mysql://admin:***@lmu-sql.cwmw1ha3dmhf.us-east-2.rds.amazonaws.com/Spotify_Data
1 rows affected.


[(Decimal('75.3300'), 100)]

In [36]:
%%sql
SELECT AVG(tracks.track_popularity) As track_popularity, COUNT(*) AS track_count
FROM tracks
JOIN audio_features
    ON tracks.track_id = audio_features.track_id
WHERE audio_features.speechiness > 
    (
    SELECT AVG(speechiness)
    FROM audio_features
    )

 * mysql://admin:***@lmu-sql.cwmw1ha3dmhf.us-east-2.rds.amazonaws.com/Spotify_Data
1 rows affected.


[(Decimal('77.2308'), 52)]

In [32]:
%%sql
SELECT AVG(tracks.track_popularity) As track_popularity, COUNT(*) AS track_count
FROM tracks
JOIN audio_features
    ON tracks.track_id = audio_features.track_id
WHERE audio_features.instrumentalness > 
    (
    SELECT AVG(instrumentalness)
    FROM audio_features
    )

 * mysql://admin:***@lmu-sql.cwmw1ha3dmhf.us-east-2.rds.amazonaws.com/Spotify_Data
1 rows affected.


[(Decimal('77.2308'), 52)]

In [34]:
%%sql
SELECT AVG(tracks.track_popularity) As track_popularity, COUNT(*) AS track_count
FROM tracks
JOIN audio_features
    ON tracks.track_id = audio_features.track_id
WHERE audio_features.liveness > 
    (
    SELECT AVG(liveness)
    FROM audio_features
    )

 * mysql://admin:***@lmu-sql.cwmw1ha3dmhf.us-east-2.rds.amazonaws.com/Spotify_Data
1 rows affected.


[(Decimal('74.5273'), 55)]

In [35]:
%%sql
SELECT AVG(tracks.track_popularity) As track_popularity, COUNT(*) AS track_count
FROM tracks
JOIN audio_features
    ON tracks.track_id = audio_features.track_id
WHERE audio_features.valence > 
    (
    SELECT AVG(valence)
    FROM audio_features
    )

 * mysql://admin:***@lmu-sql.cwmw1ha3dmhf.us-east-2.rds.amazonaws.com/Spotify_Data
1 rows affected.


[(Decimal('76.0244'), 82)]

When interpreting the results of these queries, it is important to consider the duality of track popularity and track count. Some features are high on average, but are relatively low on track count. The best features would be those who are above average on both popularity and count. Of this description, danceability, energy, and valence fit the bill the best. 

These three features have both a strong average popularity and a high count of tracks that score above average on them. In the future, Universal should pursue aspiring aritsts who they believe score high in these three respective categories.

Secondary Question: What tempo is most prevalent in popular music?

To further assist their artist scouting efforts, Universal can also equip themsleves with the knowledge of which tempo is most popular.

For this query, I will utilize a CTE to create a table comprised of the temp column and the track_id column which will then be joined to the tracks table. From there, I will use a combination of SUM and CASE to categorize each song by its tempo and count the total number of songs within each category.

In [11]:
%%sql 
WITH music_tempo AS (
    SELECT track_id, temp
    FROM audio_features
)

SELECT
SUM(CASE
    WHEN temp < 100 THEN 1
    END) AS Slow_Temp,
SUM(CASE
    WHEN temp BETWEEN 100 AND 140 THEN 1
    END) AS Med_Temp,
SUM(CASE
    WHEN temp > 140 THEN 1
    END) AS Fast_Temp
FROM tracks
JOIN music_tempo 
    ON tracks.track_id = music_tempo.track_id


 * mysql://admin:***@lmu-sql.cwmw1ha3dmhf.us-east-2.rds.amazonaws.com/Spotify_Data
1 rows affected.


[(Decimal('51'), Decimal('89'), Decimal('42'))]

As we can see, the distribution of tempos within the data falls pretty close to a normal distribution. Most of the tempos fall within the middle category, while the high and low categories each contain a fairly even smaller share of the songs. Based off this result, Universal should pursue music which tempo can be characterized as average. Their most popular songs share tempo of this variety the most.

Secondary question: Does loudness factor in to track popularity?

Spotify also attributes a rating for the loudness of each track available on the platform. Universal can attempt to interpret this figure's effect on the popularity of its artist's music.

I can look into the volume of an artists catalog by running multiple JOINS and saving them into a VIEW. This will provide easy access to the data we are most concerned with in this case. Within the VIEW, I will sum each artists total volume of their catalog through the use of a GROUP BY command. Furtermore, I will sort the results by artist popularity in order to see if the top artists have a common trend amongst their loudness attribute.

In [9]:
%%sql
CREATE OR REPLACE VIEW artist_volume AS 
    SELECT artists.artist_name, artists.artist_popularity, SUM(loudness) AS total_catalog_volume
    FROM artists
    JOIN tracks
        ON artists.artist_id = tracks.artist_id
    JOIN audio_features 
        ON tracks.track_id = audio_features.track_id
    GROUP BY artists.artist_name
    ORDER BY artist_popularity DESC

 * mysql://admin:***@lmu-sql.cwmw1ha3dmhf.us-east-2.rds.amazonaws.com/Spotify_Data
0 rows affected.


[]

In [10]:
%%sql
SELECT *
FROM artist_volume

 * mysql://admin:***@lmu-sql.cwmw1ha3dmhf.us-east-2.rds.amazonaws.com/Spotify_Data
20 rows affected.


[('Drake', 98, -60.373),
 ('Post Malone', 95, -47.464),
 ('Billie Eilish', 95, -122.67),
 ('DaBaby', 93, -31.946),
 ('Taylor Swift', 92, -62.505),
 ('Ariana Grande', 92, -53.139),
 ('Rihanna', 91, -57.687),
 ('Kanye West', 90, -45.374),
 ('Queen', 89, -79.073),
 ('Lil Wayne', 89, -50.449),
 ('Maroon 5', 89, -57.688),
 ('The Beatles', 89, -98.309),
 ('Imagine Dragons', 89, -41.462),
 ('Lady Gaga', 88, -52.49),
 ('Coldplay', 88, -61.614),
 ('Jonas Brothers', 85, -53.376),
 ('Elton John', 84, -91.133),
 ('50 Cent', 83, -42.115),
 ('Mariah Carey', 79, -71.423),
 ('Mumford & Sons', 78, -72.38)]

As we can see by the results, the total volume seems to be randomly distributed throughout the varying levels of artist popularity. From this, we can conclude that loudness of music does not seem to correlate strongly with popularity. Keeping this in mind, Universal should not give this attribute priority when looking for new artists to sign.

In [12]:
!pip install nbmerge

  Created wheel for nbmerge: filename=nbmerge-0.0.4-py2.py3-none-any.whl size=6409 sha256=0f5899f9d32c0e93e2e022e9fe7e67de2b6008e505d2ff0314bd1c525b6e386a
  Stored in directory: C:\Users\Nick\AppData\Local\pip\Cache\wheels\93\0c\59\ac91905c182ce43c4f39a50b5381c9485049cfe913237872ff
Successfully built nbmerge


In [14]:
!nbmerge final_data_collection.ipynb sql_analysis.ipynb > presentation.ipynb